src/iceberg/catalog/rest/catalog.h

```diff
@@ -0,0 +1,165 @@
+
+  /// \brief Start a transaction to create a table
+  ///
+  /// \param identifier a table identifier
+  /// \param schema a schema
+  /// \param spec a partition spec
+  /// \param location a location for the table; leave empty if unspecified
+  /// \param properties a string map of table properties
+  /// \return a Transaction to create the table or ErrorKind::kAlreadyExists if the
+  /// table already exists
+  Result<std::shared_ptr<Transaction>> StageCreateTable(
+      const TableIdentifier& identifier, const Schema& schema, const PartitionSpec& spec,
+      const std::string& location,
+      const std::unordered_map<std::string, std::string>& properties) override;
+
+  /// \brief Check whether table exists
+  ///
+  /// \param identifier a table identifier
+  /// \return Result<bool> indicating table exists or not.
+  ///         - On success, the table existence was successfully checked (actual
+  ///         existence may be inferred elsewhere).
+  ///         - On failure, contains error information.
+  Result<bool> TableExists(const TableIdentifier& identifier) const override;
+
+  /// \brief Drop a table; optionally delete data and metadata files
+  ///
+  /// If purge is set to true the implementation should delete all data and metadata
+  /// files.
+  ///
+  /// \param identifier a table identifier
+  /// \param purge if true, delete all data and metadata files in the table
+  /// \return Status indicating the outcome of the operation.
+  ///         - On success, the table was dropped (or did not exist).
+  ///         - On failure, contains error information.
+  Status DropTable(const TableIdentifier& identifier, bool purge) override;
+
+  /// \brief Load a table
+  ///
+  /// \param identifier a table identifier
+  /// \return instance of Table implementation referred to by identifier or
+  /// ErrorKind::kNoSuchTable if the table does not exist
+  Result<std::unique_ptr<Table>> LoadTable(const TableIdentifier& identifier) override;
+
+  /// \brief Register a table with the catalog if it does not exist
+  ///
+  /// \param identifier a table identifier
+  /// \param metadata_file_location the location of a metadata file
+  /// \return a Table instance or ErrorKind::kAlreadyExists if the table already exists
+  Result<std::shared_ptr<Table>> RegisterTable(
+      const TableIdentifier& identifier,
+      const std::string& metadata_file_location) override;
+
+  /// \brief A builder used to create valid tables or start create/replace transactions
+  ///
+  /// \param identifier a table identifier
+  /// \param schema a schema
+  /// \return the builder to create a table or start a create/replace transaction
+  std::unique_ptr<RestCatalog::TableBuilder> BuildTable(
+      const TableIdentifier& identifier, const Schema& schema) const override;
+
+ private:
+  RestCatalog(std::shared_ptr<RestCatalogConfig> config,
+              std::unique_ptr<HttpClient> client);
+
+  std::shared_ptr<RestCatalogConfig> config_;
```

Comment:
What is the reason that config is shared but client is unique? Shouldn't both be unique?